// @apollo/server/dist/esm/standalone/index.js
import cors from 'cors';
import express from 'express';
import { parse as parseContentType } from 'content-type';
import http from 'http';
import { expressMiddleware } from '../express4/index.js';
import { ApolloServerPluginDrainHttpServer } from '../plugin/drainHttpServer/index.js';
import { urlForHttpServer } from '../utils/urlForHttpServer.js';
const validCharset = /^utf-(8|((16|32)(le|be)?))$/i;
export async function startStandaloneServer(server, options) {
    const app = express();
    const httpServer = http.createServer(app);
    server.addPlugin(ApolloServerPluginDrainHttpServer({ httpServer: httpServer }));
    await server.start();
    const context = options?.context ?? (async () => ({}));
    app.use(cors(), express.json({
        verify(req) {
            const charset = parseContentType(req).parameters.charset || 'utf-8';
            if (!charset.match(validCharset)) {
                throw Object.assign(new Error(`unsupported charset "${charset.toUpperCase()}"`), {
                    status: 415,
                    name: 'UnsupportedMediaTypeError',
                    charset,
                    type: 'charset.unsupported',
                });
            }
        },
        limit: '50mb',
    }), expressMiddleware(server, { context }));
    const listenOptions = options?.listen ?? { port: 4000 };
    await new Promise((resolve) => {
        httpServer.listen(listenOptions, resolve);
    });
    return { url: urlForHttpServer(httpServer) };
}
